core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java

```diff
@@ -218,14 +218,14 @@ public ShardId shardId(ClusterState clusterState, String index, String id, @Null
         return new ShardId(indexMetaData.getIndex(), generateShardId(indexMetaData, id, routing));
     }
 
-    private int generateShardId(IndexMetaData indexMetaData, String id, @Nullable String routing) {
+    static int generateShardId(IndexMetaData indexMetaData, String id, @Nullable String routing) {
         final int hash;
         if (routing == null) {
             hash = Murmur3HashFunction.hash(id);
         } else {
             hash = Murmur3HashFunction.hash(routing);
         }
-        return Math.floorMod(hash, indexMetaData.getNumberOfShards());
+        return Math.floorDiv(Math.floorMod(hash, indexMetaData.getRoutingNumShards()), indexMetaData.getRoutingFactor());
```

Comment:
so lets say we have an index with `8` shards shrunk to `4`  so routing factor is `2` `Math.floorMod(hash, indexMetaData.getRoutingNumShards()) = 7` so `7/2 = 3` which is correct but it's rounded so if it is that what you mean then yes it's exact but I decided to be on the safe side... WDYT?